ReactAndroid/src/main/java/com/facebook/react/ReactDelegate.java

```diff
@@ -0,0 +1,191 @@
+package com.facebook.react;
+
+import android.app.Activity;
+import android.content.Intent;
+import android.net.Uri;
+import android.os.Build;
+import android.os.Bundle;
+import android.provider.Settings;
+import android.view.KeyEvent;
+import android.widget.Toast;
+
+import com.facebook.common.logging.FLog;
+import com.facebook.infer.annotation.Assertions;
+import com.facebook.react.common.ReactConstants;
+import com.facebook.react.devsupport.DoubleTapReloadRecognizer;
+import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler;
+
+import javax.annotation.Nullable;
+
+/**
+ * A delegate for handling React Application support. This delegate is unaware whether it is used in
+ * an {@link Activity} or a {@link android.app.Fragment}.
+ */
+public class ReactDelegate {
+
+  public final int REQUEST_OVERLAY_PERMISSION_CODE = 1111;
+  private static final String REDBOX_PERMISSION_MESSAGE =
+    "Overlay permissions need to be granted in order for react native apps to run in dev mode.";
+  private static final String REDBOX_PERMISSION_GRANTED_MESSAGE =
+    "Overlay permissions have been granted.";
+
+
+  private final Activity mActivity;
+  private ReactRootView mReactRootView;
+
+
+  @Nullable
+  private final String mMainComponentName;
+
+  @Nullable
+  private Bundle mLaunchOptions;
+
+  @Nullable
+  private DoubleTapReloadRecognizer mDoubleTapReloadRecognizer;
+
+
+  public ReactDelegate(Activity activity, String appKey, Bundle launchOptions) {
+    mActivity = activity;
+    mMainComponentName = appKey;
+    mLaunchOptions = launchOptions;
+    mDoubleTapReloadRecognizer = new DoubleTapReloadRecognizer();
+  }
+
+  public void onHostResume() {
+    if (getReactNativeHost().hasInstance()) {
+      if (mActivity instanceof DefaultHardwareBackBtnHandler) {
```

Comment:
We are currently in the process of open sourcing our solution at Hudl. Since contributing back to React Native seems to be taking longer then expected, we wanted to be able to share our approach with others. I will comment back on this PR when that happens.